src/waveform/widgets/allshader/waveformwidget.cpp

```diff
@@ -26,53 +25,70 @@ WaveformWidget::WaveformWidget(QWidget* parent,
     // The following two renderers work in tandem: if the rendered waveform is
     // for a stem track, WaveformRendererSignalBase will skip rendering and let
     // WaveformRendererStem do the rendering, and vice-versa.
-    addRenderer<WaveformRendererStem>();
+    pOpacityNode->appendChildNode(addRendererNode<WaveformRendererStem>());
 #endif
-    allshader::WaveformRendererSignalBase* waveformSignalRenderer =
-            addWaveformSignalRenderer(
-                    type, options, ::WaveformRendererAbstract::Play);
-
-    addRenderer<WaveformRenderBeat>();
-    addRenderer<WaveformRenderMark>();
-
-    // if the signal renderer supports slip, we add it again, now for slip, together with the
-    // other slip renderers
-    if (waveformSignalRenderer && waveformSignalRenderer->supportsSlip()) {
+    pOpacityNode->appendChildNode(addWaveformSignalRendererNode(
+            type, options, ::WaveformRendererAbstract::Play));
+    pOpacityNode->appendChildNode(addRendererNode<WaveformRenderBeat>());
+    pOpacityNode->appendChildNode(addRendererNode<WaveformRenderMark>());
+    m_pWaveformRenderMark = static_cast<WaveformRenderMark*>(pOpacityNode->lastChild());
+
+    // if the added signal renderer supports slip, we add it again, now for
+    // slip, together with the other slip renderers
+    if (m_pWaveformRendererSignal && m_pWaveformRendererSignal->supportsSlip()) {
         // The following renderer will add an overlay waveform if a slip is in progress
-        addRenderer<WaveformRendererSlipMode>();
-        addRenderer<WaveformRendererPreroll>(::WaveformRendererAbstract::Slip);
+        pOpacityNode->appendChildNode(addRendererNode<WaveformRendererSlipMode>());
+        pOpacityNode->appendChildNode(
+                addRendererNode<WaveformRendererPreroll>(
+                        ::WaveformRendererAbstract::Slip));
 #ifdef __STEM__
-        addRenderer<WaveformRendererStem>(::WaveformRendererAbstract::Slip);
+        pOpacityNode->appendChildNode(
+                addRendererNode<WaveformRendererStem>(
+                        ::WaveformRendererAbstract::Slip));
 #endif
-        addWaveformSignalRenderer(type, options, ::WaveformRendererAbstract::Slip);
-        addRenderer<WaveformRenderBeat>(::WaveformRendererAbstract::Slip);
-        addRenderer<WaveformRenderMark>(::WaveformRendererAbstract::Slip);
+        pOpacityNode->appendChildNode(addWaveformSignalRendererNode(
+                type, options, ::WaveformRendererAbstract::Slip));
+        pOpacityNode->appendChildNode(
+                addRendererNode<WaveformRenderBeat>(
+                        ::WaveformRendererAbstract::Slip));
+        pOpacityNode->appendChildNode(
+                addRendererNode<WaveformRenderMark>(
+                        ::WaveformRendererAbstract::Slip));
     }
 
     m_initSuccess = init();
+
+    pTopNode->appendChildNode(std::move(pOpacityNode));
+    m_pOpacityNode = static_cast<rendergraph::OpacityNode*>(pTopNode->lastChild());
```

Comment:
I see this seems to be a frequent usecase, does this warrant an API change so something like this is possible?
```suggestion
    m_pOpacityNode = pTopNode->appendChildNode(std::move(pOpacityNode));
```

That would be fairly straightforward with a template like this:
```c++
template <typename T>
T* Node::appendChildNode(std::unique_ptr<T> ptr) {
    // call the non-template type-erased code here
    // ... and then return the leaked pointer in case the caller needs it
    return ptr.get();
}
```
This would avoid issues where the two lines might get detached from each other (or `lastChild` just doesn't return the right object for some reason) and the `static_cast` goes wrong. This API seems much safer than the current solution without much extra complexity. Wdyt?

Reply:
Very nice and elegant! I will make the change!

Reply:
done